clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1889,24 +1893,27 @@ static void printArgument(ASTContext &Ctx, raw_ostream &ArgOS,
     TypePolicy.SuppressTagKeyword = true;
     QualType T = Arg.getAsType();
     QualType FullyQualifiedType = TypeName::getFullyQualifiedType(T, Ctx, true);
+    if (!FirstArg)
+      ArgOS << ", ";
     ArgOS << FullyQualifiedType.getAsString(TypePolicy);
     break;
   }
   default:
+    if (!FirstArg)
+      ArgOS << ", ";
     Arg.print(P, ArgOS);
   }
 }
 
 static void printArguments(ASTContext &Ctx, raw_ostream &ArgOS,
                            ArrayRef<TemplateArgument> Args,
                            const PrintingPolicy &P) {
+  bool FirstArg = true;
+
   for (unsigned I = 0; I < Args.size(); I++) {
     const TemplateArgument &Arg = Args[I];
-
-    if (I != 0)
```

Comment:
I'd probably leave this in place, and just add a 'if (Arg.isPackExpansion() && Arg.pack_size() ==0) contine;"

Reply:
Also, check out all the values for getKind() here: https://clang.llvm.org/doxygen/TemplateBase_8cpp_source.html#l00171

It seems we don't handle nearly enough of those...And interestingly, the isPackExpansion doesn't include 'Pack'.  So perhaps the first part of my suggestion above isn't correct, but I still think that should be here..

Alternatively, you could print the comma AFTER printArgument only if printArguments returns 'true' (in that it printed something) AND I != Args.size() -1.

That is probably the cleanest way to do this.

Reply:
You have to look ahead. Consider lit test below - the arguments are 'char' and 'empty pack'. printArgument will print char successfully but the comma should not be printed because this will result in <char, >. Basically the we need to print  <comma>arg as opposed to arg<comma>. So I moved comma handling to printArgument because this is where argument type is checked and printed.  I guess I could check the type of the next argument in printArguments if you prefer that.

Reply:
Hmm... right.  I'm just finding that the mechanism seems clunkly.  I like the idea of having this function manage the commas (since it is a 'join' function), and I like having the printArgument have the logic, and return true/false, but I can't really think of a good way to make this.  

I think then unless @premanandrao  or @Fznamznon makes a better suggestion, that we should just leak the 'arg pack' test into this function and put it where you checked I !=0 above.

Reply:
I understand. I liked it separated out too but I couldn't figure out a good way to do it. I'll wait a bit to see if someone else has a better suggestion, but if not I'll add the 'pack check' for the next argument in printArguments